### PR TITLE
Refactor graceful manager, fix misused WaitGroup

### PR DESCRIPTION
**Pull request by forsaken628**
Follow #29629

### CODE REVIEW

**Comment by forsaken628** [comment]
I have an off topic question. Why was `http.Server.Shutdown` not used for http graceful shutdown ?

File: modules/graceful/manager_unix.go [hunk 1]
```diff
@@ -57,37 +57,35 @@ func (g *Manager) start() {
...
+				g.notify(readyMsg)
+				return
+			}
+			select {
+			case <-g.IsShutdown():
+				g.createServerCond.L.Unlock()
+				return
+			default:
+			}
+			g.createServerCond.Wait()
```
Comment (wxiaoguang) [inline modules/graceful/manager_unix.go:79]: Is it right? When `Wait`, the lock `L` is acquired here. Then the `L.Lock()` in InformCleanup would deadlock?

Comment (forsaken628) [reply modules/graceful/manager_unix.go:79]:
```
Wait atomically unlocks c.L and suspends execution of the calling goroutine. After later resuming execution, Wait locks c.L before returning. Unlike in other systems, Wait cannot return unless awoken by Broadcast or Signal. 
```
https://pkg.go.dev/sync@go1.22.1#Cond.Wait

Comment (wxiaoguang) [reply modules/graceful/manager_unix.go:79]: Oops, I haven't used sync.Cond for long time and forgot the details. Sorry for the misread.

**Review by wxiaoguang — APPROVED** [review]

File: modules/graceful/manager_unix.go [hunk 2]
```diff
@@ -57,37 +57,35 @@ func (g *Manager) start() {
...
-			g.createServerWaitGroup.Wait()
+		defer func() {
+			close(startupDone)
+			// Ignore the error here there's not much we can do with it, they're logged in the CloseProvidedListeners function
+			_ = CloseProvidedListeners()
 		}()
-		// Ignore the error here there's not much we can do with it, they're logged in the CloseProvidedListeners function
-		_ = CloseProvidedListeners()
-		g.notify(readyMsg)
+		// Wait till we're done getting all the listeners and then close the unused ones
```
Comment (nanguanlin6) [inline modules/graceful/manager_unix.go]: Can this comment be removed or rephrased like `Wait for all servers to be created`?

File: modules/graceful/manager_unix.go [hunk 3]
```diff
@@ -57,37 +57,35 @@ func (g *Manager) start() {
...
-			// If nothing has started, the "Wait" just panics: sync: WaitGroup is reused before previous Wait has returned
-			// There is no clear solution besides a complete rewriting of the "manager"
-			defer func() {
-				_ = recover()
-			}()
-			g.createServerWaitGroup.Wait()
+		defer func() {
+			close(startupDone)
+			// Ignore the error here there's not much we can do with it, they're logged in the CloseProvidedListeners function
+			_ = CloseProvidedListeners()
```
Comment (nanguanlin6) [inline modules/graceful/manager_unix.go:63]: And there add something like `close the unused ones`

**Review by nanguanlin6 — APPROVED** [review]

File: modules/graceful/manager_unix.go [hunk 4]
```diff
@@ -57,37 +57,35 @@ func (g *Manager) start() {
...
+			// Close the unused listeners and ignore the error here there's not much we can do with it, they're logged in the CloseProvidedListeners function
+			_ = CloseProvidedListeners()
 		}()
-		// Ignore the error here there's not much we can do with it, they're logged in the CloseProvidedListeners function
-		_ = CloseProvidedListeners()
-		g.notify(readyMsg)
+		// Wait for all servers to be created
+		g.createServerCond.L.Lock()
+		for {
+			if g.createdServer >= numberOfServersToCreate {
```
Comment (lunny) [inline modules/graceful/manager_unix.go:68]: When and Where `g.createdServer++` will be invoked? I couldn't find it.

**Comment by lunny** [comment]
> I have an off topic question. Why was `http.Server.Shutdown` not used for http graceful shutdown ?

I think maybe when Gitea start developing, the functions on std library haven't been introduced.

**Pull request merged**